File: certification-management/examples/nodejs/tls-demo.js
const https = require("https");
const fs = require("fs");
const path = require("path");

const PKI_DIR = path.resolve(__dirname, "../../pki");

function createTLSServer(port = 8443) {
  const options = {
    key: fs.readFileSync(
      path.join(PKI_DIR, "certs/server/server.myapp.local.key")
    ),
    cert: fs.readFileSync(
      path.join(PKI_DIR, "certs/server/server.myapp.local-fullchain.crt")
    ),
    minVersion: "TLSv1.2",
    ciphers: "ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20",
    honorCipherOrder: true,
  };

  const server = https.createServer(options, (req, res) => {
    res.writeHead(200, { "Content-Type": "text/plain" });
    res.end("Hello, TLS from Node.js!\n");
  });

  server.listen(port, () => {
    console.log(`TLS server listening on https://localhost:${port}`);
  });

  return server;
}

function createTLSClient(host = "server.myapp.local", port = 8443) {
  const ca = fs.readFileSync(path.join(PKI_DIR, "root-ca/ca.crt"));

  const options = {
    hostname: host,
    port: port,
    path: "/",
    method: "GET",
    ca: ca,
    minVersion: "TLSv1.2",
    rejectUnauthorized: true,
  };

  const req = https.request(options, (res) => {
    let data = "";
    res.on("data", (chunk) => {
      data += chunk;
    });
    res.on("end", () => {
      console.log(`Response: ${data}`);
      console.log(`TLS Version: ${res.socket.getProtocol?.() || "N/A"}`);
      console.log(
        `Cipher: ${res.socket.getCipher?.()?.name || "N/A"}`
      );
    });
  });

  req.on("error", (e) => {
    console.error(`Request error: ${e.message}`);
  });

  req.end();
}

function createMTLSServer(port = 8444) {
  const options = {
    key: fs.readFileSync(
      path.join(PKI_DIR, "certs/server/server.myapp.local.key")
    ),
    cert: fs.readFileSync(
      path.join(PKI_DIR, "certs/server/server.myapp.local-fullchain.crt")
    ),
    ca: fs.readFileSync(path.join(PKI_DIR, "root-ca/ca.crt")),
    requestCert: true,
    rejectUnauthorized: true,
    minVersion: "TLSv1.2",
    ciphers: "ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20",
  };

  const server = https.createServer(options, (req, res) => {
    const clientCert = req.socket.getPeerCertificate();
    res.writeHead(200, { "Content-Type": "text/plain" });
    res.end(
      `Hello, mTLS! Client CN: ${clientCert.subject?.CN || "Unknown"}\n`
    );
  });

  server.listen(port, () => {
    console.log(`mTLS server listening on https://localhost:${port}`);
  });

  return server;
}

function createMTLSClient(host = "server.myapp.local", port = 8444) {
  const options = {
    hostname: host,
    port: port,
    path: "/",
    method: "GET",
    ca: fs.readFileSync(path.join(PKI_DIR, "root-ca/ca.crt")),
    key: fs.readFileSync(
      path.join(PKI_DIR, "certs/client/client.myapp.local.key")
    ),
    cert: fs.readFileSync(
      path.join(PKI_DIR, "certs/client/client.myapp.local.crt")
    ),
    minVersion: "TLSv1.2",
    rejectUnauthorized: true,
  };

  const req = https.request(options, (res) => {
    let data = "";
    res.on("data", (chunk) => {
      data += chunk;
    });
    res.on("end", () => {
      console.log(`Response: ${data}`);
    });
  });

  req.on("error", (e) => {
    console.error(`Request error: ${e.message}`);
  });

  req.end();
}

const command = process.argv[2];

switch (command) {
  case "server":
    createTLSServer();
    break;
  case "client":
    createTLSClient();
    break;
  case "mtls-server":
    createMTLSServer();
    break;
  case "mtls-client":
    createMTLSClient();
    break;
  default:
    console.log(
      "Usage: node tls-demo.js [server|client|mtls-server|mtls-client]"
    );
    process.exit(1);
}

File: certification-management/examples/python/tls_demo.py
import ssl
import socket
import subprocess
import sys
from pathlib import Path

PKI_DIR = Path(__file__).parent.parent.parent / "pki"


def create_tls_server(host="0.0.0.0", port=8443):
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
    context.minimum_version = ssl.TLSVersion.TLSv1_2
    context.load_cert_chain(
        certfile=PKI_DIR / "certs" / "server" / "server.myapp.local-fullchain.crt",
        keyfile=PKI_DIR / "certs" / "server" / "server.myapp.local.key",
    )
    context.set_ciphers(
        "ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20"
    )
    context.options |= ssl.OP_NO_SSLv2
    context.options |= ssl.OP_NO_SSLv3
    context.options |= ssl.OP_NO_TLSv1
    context.options |= ssl.OP_NO_TLSv1_1

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.bind((host, port))
        sock.listen(5)
        print(f"TLS server listening on {host}:{port}")

        with context.wrap_socket(sock, server_side=True) as ssock:
            conn, addr = ssock.accept()
            print(f"Connection from {addr}")
            print(f"TLS Version: {conn.version()}")
            print(f"Cipher: {conn.cipher()}")
            data = conn.recv(1024)
            conn.sendall(b"Hello, TLS! " + data)


def create_tls_client(host="server.myapp.local", port=8443):
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    context.minimum_version = ssl.TLSVersion.TLSv1_2
    context.load_verify_locations(PKI_DIR / "root-ca" / "ca.crt")
    context.set_ciphers(
        "ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20"
    )

    with socket.create_connection((host, port)) as sock:
        with context.wrap_socket(sock, server_hostname=host) as ssock:
            print(f"TLS Version: {ssock.version()}")
            print(f"Cipher: {ssock.cipher()}")
            ssock.sendall(b"Hello from Python client!")
            data = ssock.recv(1024)
            print(f"Received: {data.decode()}")


def create_mtls_server(host="0.0.0.0", port=8444):
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
    context.minimum_version = ssl.TLSVersion.TLSv1_2
    context.load_cert_chain(
        certfile=PKI_DIR / "certs" / "server" / "server.myapp.local-fullchain.crt",
        keyfile=PKI_DIR / "certs" / "server" / "server.myapp.local.key",
    )
    context.load_verify_locations(PKI_DIR / "root-ca" / "ca.crt")
    context.verify_mode = ssl.CERT_REQUIRED

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.bind((host, port))
        sock.listen(5)
        print(f"mTLS server listening on {host}:{port}")

        with context.wrap_socket(sock, server_side=True) as ssock:
            conn, addr = ssock.accept()
            print(f"Connection from {addr}")
            print(f"Client CN: {conn.getpeercert()['subject']}")
            data = conn.recv(1024)
            conn.sendall(b"Hello, mTLS! " + data)


def create_mtls_client(host="server.myapp.local", port=8444):
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    context.minimum_version = ssl.TLSVersion.TLSv1_2
    context.load_verify_locations(PKI_DIR / "root-ca" / "ca.crt")
    context.load_cert_chain(
        certfile=PKI_DIR / "certs" / "client" / "client.myapp.local.crt",
        keyfile=PKI_DIR / "certs" / "client" / "client.myapp.local.key",
    )

    with socket.create_connection((host, port)) as sock:
        with context.wrap_socket(sock, server_hostname=host) as ssock:
            print(f"TLS Version: {ssock.version()}")
            print(f"Cipher: {ssock.cipher()}")
            ssock.sendall(b"Hello from Python mTLS client!")
            data = ssock.recv(1024)
            print(f"Received: {data.decode()}")


def check_cert_expiry(domain, port=443):
    context = ssl.create_default_context()
    with socket.create_connection((domain, port)) as sock:
        with context.wrap_socket(sock, server_hostname=domain) as ssock:
            cert = ssock.getpeercert()
            print(f"Domain: {domain}")
            print(f"Subject: {cert['subject']}")
            print(f"Issuer: {cert['issuer']}")
            print(f"Not Before: {cert['notBefore']}")
            print(f"Not After: {cert['notAfter']}")
            print(f"SAN: {cert.get('subjectAltName', 'N/A')}")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python tls_demo.py [server|client|mtls-server|mtls-client|check <domain>]")
        sys.exit(1)

    cmd = sys.argv[1]
    if cmd == "server":
        create_tls_server()
    elif cmd == "client":
        create_tls_client()
    elif cmd == "mtls-server":
        create_mtls_server()
    elif cmd == "mtls-client":
        create_mtls_client()
    elif cmd == "check":
        domain = sys.argv[2] if len(sys.argv) > 2 else "google.com"
        check_cert_expiry(domain)
    else:
        print(f"Unknown command: {cmd}")
